### PR TITLE
Downport: 1.20.1

### PR DESCRIPTION
**Pull request by MianReplicate**
This PR aims to port the current master branch 1.20.2 version of TR to 1.20.1 in full working state for both Neo (hopefully) and Fabric.

Can I just say that I am suffering? This port forced me to look into what a codec was to fix one (and I am 99% sure it's not fixed!)

### CODE REVIEW

**Comment by Olivier1152005** [comment]
Is there an easy way to download the decompiled version of the mod that is on GitHub without having to manually go to each file, download it, and then manually rebuild every file and path? Is there anything in particular you are having trouble with? I like this mod and want to help get it ported to 1.20.1.

**Comment by CommandrMoose** [comment]
> Is there an easy way to download the decompiled version of the mod that is on GitHub without having to manually go to each file, download it, and then manually rebuild every file and path? Is there anything in particular you are having trouble with? I like this mod and want to help get it ported to 1.20.1.

I recommend looking into how git works. This PR is close to coming to a close.
https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository

**Comment by Jeryn99** [comment]
> Is there an easy way to download the decompiled version of the mod that is on GitHub without having to manually go to each file, download it, and then manually rebuild every file and path? Is there anything in particular you are having trouble with? I like this mod and want to help get it ported to 1.20.1.

Hello. What you are viewing is code that is cutting edge. What I mean by that is that it is so newly published that if you were to cut the review and dev process out and ran off with these files, you would end up destroying your Minecraft saves. This is a pull request for developers for tracking the port downwards and if you do not know what you are doing, you should not interact as you are hindering the process. Think of it as walking into a car factory and asking to take the car in its current state (Engine, no brakes yet though) because you love it, just wouldn't go down well

**Comment by 50ap5ud5** [comment]
PR source branch needs to be switched to a recently created branch called [minecraft/1.20.1_port](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1_port) before being moved out of a draft state.

It is a copy of the porting work but rebased on top of the latest 20.2 changes. This means this PR would up to date with the latest 1.20.2 changes. Those reviewing the PR would only be seeing changes that were made during the port.

**Comment by Jeryn99** [comment]
This does sound like we need to make a whole new PR?

**Comment by 50ap5ud5** [comment]
> This does sound like we need to make a whole new PR?

Actually, I think you are right here. 

I had made a new branch in case I significantly changed the commit history. It looks it wasn't the case, so if we want to continue using this PR, we just need to override the [LoxiGoose:minecraft/1.20.1](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1) branch with that of the [LoxiGoose:minecraft/1.20.1_port](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1_port) branch.

**Comment by 50ap5ud5** [comment]
Ok, I've updated [LoxiGoose:minecraft/1.20.1](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1) to be rebased on top of [WhoCraft:minecraft/1.20](https://github.com/WhoCraft/TardisRefined/tree/minecraft/1.20) changes so that we don't need to make a new PR.

**Comment by Jeryn99** [comment]
we're really damn cooking

File: common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java [hunk 1]
```diff
@@ -15,12 +17,7 @@ public abstract class DeferredRegistry<T> {
     /** Register using a Supplier */
     public abstract <R extends T> RegistrySupplier<R> register(String id, Supplier<R> supplier);
 
-    /** Create a RegistrySupplierHolder. This has similar behaviour to a Supplier but has vanilla Holder attributes such as use in tags*/
-    public abstract <I extends T> RegistrySupplierHolder<T, I> registerHolder(final String name, final Supplier<I> sup);
```
Comment (50ap5ud5) [inline common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java:82]: Why was the ability to register via a Holder removed? The reason it was added in 20.2 was because a ``Holder`` is a vanilla equivalent of our ``RegistrySupplier``. I had based our version off Neoforge's DeferredHolder to allow for us to start to move towards using more vanilla systems so that it will be easier to detect if issues came from us or vanilla's upstream logic.

File: common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java [hunk 2]
```diff
@@ -15,12 +17,7 @@ public abstract class DeferredRegistry<T> {
...
-
-    /** Get the underlying registry, which includes all entries added by any mod that has a DeferredRegistry with the same ResourceKey.
-     * All lookup methods should be called from here.*/
-    public abstract Registry<T> getRegistry();
```
Comment (50ap5ud5) [inline common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java]: Why was the ability to get the underlying Registry removed? This was added in 20.2 so that we can utilise methods specific to the vanilla `Registry<T>` object.

File: common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java [hunk 3]
```diff
@@ -34,19 +31,7 @@ public static <T> DeferredRegistry<T> create(String modid, ResourceKey<? extends
         throw new RuntimeException(TardisRefined.PLATFORM_ERROR);
     }
 
-    /** Gets the underlying Codec for the registry object type, if defined. Currently unused and untested, but added for completeness*/
-    public abstract Supplier<Codec<T>> getCodec();
```
Comment (50ap5ud5) [inline common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java]: Why was this limited to only the `CustomRegistry`? `getCodec` was added in 20.2 for all of our mod's registries because many registries in vanilla use an underlying codec, such as BlockStates and ItemStacks.

If in the future we needed to create any data-driven features this method would have allowed us to retrieve the codec to use for data read/writing.

File: common/src/main/java/whocraft/tardis_refined/client/screen/selections/ShellSelectionScreen.java [hunk 4]
```diff
@@ -152,10 +154,10 @@ public void render(GuiGraphics guiGraphics, int i, int j, float f) {
         super.render(guiGraphics, i, j, f);
     }
 
-    @Override
-    public void renderBackground(GuiGraphics guiGraphics, int i, int j, float f) {
-        // super.renderBackground(guiGraphics, i, j, f);
-    }
+//    @Override
```
Comment (CommandrMoose) [inline common/src/main/java/whocraft/tardis_refined/client/screen/selections/ShellSelectionScreen.java]: Remove comments

File: forge/src/main/java/whocraft/tardis_refined/registry/forge/CustomRegistryImpl.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    private final boolean syncToClient;
+
+    public CustomRegistryImpl(ResourceLocation id, Supplier<IForgeRegistry<T>> parent, boolean syncToClient) {
+        this.parent = parent;
+        this.resourceKey = ResourceKey.createRegistryKey(id);
+        this.syncToClient = syncToClient;
+    }
+
+    public static <T> CustomRegistry<T> create(ResourceLocation id, boolean syncToClient) {
+        DeferredRegister<T> deferredRegister = DeferredRegister.create(id, id.getNamespace());
```
Comment (50ap5ud5) [inline forge/src/main/java/whocraft/tardis_refined/registry/forge/CustomRegistryImpl.java]: This implementation looks to be quite different from upstream 20.2. Any reason for this refactor?

Here `CustomRegistry` is creating a new Forge `DeferredRegister` instance, and later in `DeferredRegistryImpl`, we are creating a duplicate Forge `DeferredRegister` by calling `DeferredRegister#create` a second time.

Ideally we should have one `DeferredRegistry` instance, so that it simplifies the code for addon mod developers, any lookup operations (`getKeys`, `getValues`) would be looking from a single source of truth.
Additionally, it would reduce developer confusion when deciding whether to use `CustomRegistry` or `DeferredRegistry` for lookup values, and reduce maintenance efforts between 20.1 and 20.2.

File: fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java [hunk 6]
```diff
@@ -22,35 +24,19 @@ public static <T> DeferredRegistry<T> create(String modid, ResourceKey<? extends
...
-
     @SuppressWarnings({"unchecked", "ConstantConditions"})
     public static class Impl<T> extends DeferredRegistry<T> {
 
         private final String modid;
         private final WritableRegistry<T> registry;
-        private final boolean isCustom;
-        private final boolean syncToClient;
         private ResourceKey<? extends Registry<T>> registryKey;
+        private final List<RegistrySupplier<T>> entries;
```
Comment (50ap5ud5) [inline fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java]: `getEntries` should not be looking up from a separate list of entries. We need to be using the source of truth, which is the underlying `WritableRegistry<T>` for any lookups, because only that registry will contain any new additions added by addon mods.

File: common/src/main/java/whocraft/tardis_refined/registry/RegistrySupplierHolder.java [hunk 7]
```diff
@@ -1,228 +0,0 @@
...
-
-/**
- * An implementation of vanilla's Holder where it is a Holder constructed only with a ResourceKey.
- * It will be populated with the underlying Holder from the registry when available.
- * @param <T> The type of object being held by this RegistryHolder
- *
- * <br> Based off NeoForge's DeferredHolder
- * <br> https://github.com/neoforged/NeoForge/blob/1.20.x/src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java
- * */
-public class RegistrySupplierHolder<R, T extends R> implements Holder<R>, Supplier<T> {
```
Comment (50ap5ud5) [inline common/src/main/java/whocraft/tardis_refined/registry/RegistrySupplierHolder.java]: Why was this removed? It was kept as an alternative way to register objects which need to use vanilla features such as the ability to specify in datapack tags.

File: fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java [hunk 8]
```diff
@@ -62,25 +48,13 @@ public void register() {
...
-        @Override
-        public Registry<T> getRegistry() {
-            return this.registry;
-        }
-
-        @Override
-        public Supplier<Codec<T>> getCodec() {
-            return () -> this.registry.byNameCodec();
+        public List<RegistrySupplier<T>> getEntries() {
+            return this.entries;
```
Comment (50ap5ud5) [inline fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java]: Implementing these methods is good for convenience purposes, but it needs to call the relevant methods from the underlying `Registry<T>`. It's needed to ensure that we are looking at the list of entries which addon mods have the ability to add to.

File: forge/src/main/java/whocraft/tardis_refined/registry/forge/DeferredRegistryImpl.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    @SuppressWarnings("unchecked")
+    public static class Impl<T> extends DeferredRegistry<T> {
+
+        private final DeferredRegister<T> register;
+
+        private Registry<T> registry;
+
+        private ResourceKey<? extends Registry<T>> registryKey;
+        private final List<RegistrySupplier<T>> entries;
```
Comment (50ap5ud5) [inline forge/src/main/java/whocraft/tardis_refined/registry/forge/DeferredRegistryImpl.java]: `getEntries` should not be looking up from a separate list of entries. We need to be using the source of truth, which is the underlying `WritableRegistry<T>` for any lookups, because only that registry will contain any new additions added by addon mods.

File: forge/src/main/java/whocraft/tardis_refined/registry/forge/DeferredRegistryImpl.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        public <R extends T> RegistrySupplier<R> register(String id, Supplier<R> supplier) {
+            var orig = this.register.register(id, supplier);
+            var registrySupplier = new RegistrySupplier<>(orig.getId(), orig);
+            this.entries.add((RegistrySupplier<T>) registrySupplier);
+            return registrySupplier;
+        }
+
+        @Override
+        public List<RegistrySupplier<T>> getEntries() {
+            return this.entries;
```
Comment (50ap5ud5) [inline forge/src/main/java/whocraft/tardis_refined/registry/forge/DeferredRegistryImpl.java]: Implementing these methods is good for convenience purposes, but it needs to call the relevant methods from the underlying `Registry<T>`. It's needed to ensure that we are looking at the list of entries which addon mods have the ability to add to.

File: common/src/main/java/whocraft/tardis_refined/client/screen/upgrades/UpgradeWidget.java [hunk 11]
```diff
@@ -194,68 +196,59 @@ public void drawHover(GuiGraphics guiGraphics, int i, int j, float f, int k, int
...
+                guiGraphics.blitNineSliced(UpgradesScreen.WIDGETS,  m, l, this.width, n, 10, 200, 26, 0, 52);
             }
         }
 
-        guiGraphics.blitSprite(UpgradesScreen.getBox(upgradeEntry.isUnlocked(upgradeHandler)), 200, 26, 0, 0, q, p, n, 26);
-        guiGraphics.blitSprite(UpgradesScreen.getBox(upgradeEntry.isUnlocked(upgradeHandler)), 200, 26, 200 - o, 0, q + n, p, o, 26);
-        guiGraphics.blitSprite(UpgradesScreen.getFrame(upgradeEntry.getUpgradeType(), upgradeEntry.isUnlocked(upgradeHandler)), i + this.x + 3, j + this.y, 26, 26);
+        //blit(ResourceLocation atlasLocation, int x, int y, float uOffset, float vOffset, int width, int height, int textureWidth, int textureHeight)
+
+      //  guiGraphics.blit(UpgradesScreen.getBox(upgradeEntry.isUnlocked(upgradeHandler)), i + this.x + 3, j + this.y, 0,0, 200, 26,200, 26);
```
Comment (CommandrMoose) [inline common/src/main/java/whocraft/tardis_refined/client/screen/upgrades/UpgradeWidget.java:221]: Remove comments

File: common/src/main/java/whocraft/tardis_refined/client/screen/waypoints/WaypointListScreen.java [hunk 12]
```diff
@@ -153,18 +154,18 @@ public GenericMonitorSelectionList createSelectionList() {
     }
 
 
-    @Override
-    public void renderBackground(GuiGraphics guiGraphics, int i, int j, float f) {
-        // super.renderBackground(guiGraphics, i, j, f);
-    }
+//    @Override
+//    public void renderBackground(GuiGraphics guiGraphics, int i, int j, float f) {
```
Comment (CommandrMoose) [inline common/src/main/java/whocraft/tardis_refined/client/screen/waypoints/WaypointListScreen.java]: Going to write it for the final time here. There's lots around the place

**Review by CommandrMoose — CHANGES_REQUESTED** [review]
Comments either need to be removed or please tell my why they should stay.

There's a few changes on 1.20.2's branch that should be cherry picked in, but then I'd say it is GTG

File: common/src/main/java/whocraft/tardis_refined/registry/CustomRegistry.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+public abstract class CustomRegistry<T> {
```
Comment (50ap5ud5) [inline common/src/main/java/whocraft/tardis_refined/registry/CustomRegistry.java]: This implementation looks to be quite different from upstream 20.2. Any reason for this refactor?

**Review by 50ap5ud5 — CHANGES_REQUESTED** [review]
This is amazing work. The refactors made in the registry classes need changes as they currently prevent the ability for add-on mods to add new entries to TARDIS Refined. It also appears to remove previous functionality, some of which can be addressed with some better documentation.

I have incorporated some of your changes into the upstream 20.2 branch so that it is closer to this implementation, but generally it would be easier for maintenance if we kept similar implementation across versions where possible.

**Comment by Jeryn99** [comment]
> This is amazing work. The refactors made in the registry classes need changes as they currently prevent the ability for add-on mods to add new entries to TARDIS Refined. It also appears to remove previous functionality, some of which can be addressed with some better documentation.

When getting 1.20.1 running, we ripped out all the registry stuff in favour of CustomRegistry from Regeneration/Weeping Angels as a temporary measure as the intention was to get the mod running and see what we would need to fix at runtime. CustomRegistry was always a temporary measure and I should have documented it as so. The intention post fixes was to have your registry refactor ported down

**Comment by Jeryn99** [comment]
> Comments either need to be removed or please tell my why they should stay.
> 
> There's a few changes on 1.20.2's branch that should be cherry picked in, but then I'd say it is GTG

All commented code has now been removed from the mod where not needed

image

File: common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java (hunk 1 above)
Comment (50ap5ud5) [reply common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java:82]: Forge's registry system in 1.20.1 still retains an older design which was completely isolated from vanilla. The inability to access the underlying vanilla registry and vanilla objects such as Holders make this unfeasible unfortunately.

File: common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java (hunk 2 above)
Comment (50ap5ud5) [reply common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java]: Forge's registry system in 1.20.1 still retains an older design which was completely isolated from vanilla. The inability to access the underlying vanilla registry make this method unfeasible unfortunately. Instead, we need to expose lookup methods such as keySet and entrySet manually, so that a common interface is present for both Fabric and Forge experiences.

File: fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java (hunk 6 above)
Comment (50ap5ud5) [reply fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java]: Fixed by downporting 1.20.1 registry systems which use the underlying registry as the source of truth

**Review by 50ap5ud5 — APPROVED** [review]
